README.md
# FIELDAQ

## Overview and Resources
The software for the FIELDAQ is written in python, using the kivy library. Versioning is managed using git via GitHub. The layout and functionality are based on the old DARLING software. With the addition of a touchscreen, kivy was utilized over pygame.

- [Resources for Python](https://www.python.org/about/gettingstarted/)
- Resources for kivy can be found in [Documentation/Kivy Training Resources](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Kivy%20Training%20Resources.docx)
  - OR [online](https://kivy.org/doc/stable/gettingstarted/index.html)
- [Resources for git](https://guides.github.com/introduction/git-handbook/)
  - [The old DARLING repository](https://github.com/byu-crop-biomechanics-lab/DARLING_Software.git)

## Running the Software
After installing python and the nessesary libraries, the software is run by navigating into the [FIELDAQ/Granusoft/src](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/tree/master/Granusoft/src) directory and running the [main.py](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Granusoft/src/main.py) file.

The [main.py](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Granusoft/src/main.py) file imports the necessary kivy library elements, and builds the kivy app. The companion [main.kv](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Granusoft/src/main.kv) file imports the widgets (GUI screen elements), created for the use across the GUI, and imports the views that will be displayed throughout the app. These views are each composed of a python file and kivy file. More detail can be found in [Documentation/Kivy Screen Management.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Kivy%20Screen%20Management.docx).

The software is designed to be run on the RaspberryPi, but can also be run on any computer with python3 and kivy. This approach is useful for development and testing purposes. The code is structured to use false sensor input data when the sensor connections can not be properly detected.

## Documentation Directory
The Documentation directory contains many useful files, some of which have been mentioned previously. Documentation that has been found to be the most useful is listed here:
- [Setup Software for Non-RPi Computers.txt](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Setup%20Software%20for%20Non-RPi%20Computers.txt) -  Information on how to get the GUI running on a non-RaspberryPi computer for software development.
- [Software Update Procedure.md](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Software%20Update%20Procedure.md) - Information on how to utilize a USB drive or Github when updating the on-device software.
- [FIELDAQ_Software_Diagrams.pdf](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/FIELDAQ_Software_Diagrams.pdf) - Information about the screen navigation for the GUI and how each screen is built using kivy elements/widgets (flowchart is outdated).
- [Example_Adding the Camera Module.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Example_Adding%20the%20Camera%20Module.docx) - A reflection on how the camera screens were added.
- [Example_Adding a Boot Screen.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Example_Adding%20a%20Boot%20Screen.docx) - A hypothetical situation where a boot screen is created to choose between the Camera test mode and the Stalk Push testing mode.


Additionaly, a brief overview of all the documentation is provide here:
### Structural Information of the Code Base (how is it assembled)
- [Directory Tree.txt](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Directory%20Tree.txt) -  Overview of the directories in the repository. With information about the contents of each directory.
- [FIELDAQ_Software_Diagrams.pdf](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/FIELDAQ_Software_Diagrams.pdf) - 	Information about the screen navigation for the GUI and how each screen is built using kivy elements/widgets (flowchart is outdated).
- [Software Flowchart and Logic.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Software%20Flowchart%20and%20Logic.docx) - 	Supplementary to FIELDAQ_Software_Diagrams, this document contains more information about the purpose of each screen.
- [Software Documentation Outdated.pdf](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Software%20Documentation%20Outdated.pdf) - 	An old resource containing the Classes and Methods contained in every file. This was last updated at the end of capstone, and may contain some useful information still. An updated version can be compiled using [DoxyGen](https://www.doxygen.nl/index.html) (version. 1.8.15).
### Background Information (how do the pieces work)
- [Kivy Screen Management.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Kivy%20Screen%20Management.docx) - 	Information on how kivy navigates between screens and how screens can be added to the GUI.
- [Kivy Training Resources.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Kivy%20Training%20Resources.docx) - 	A list of resources on how to learn kivy for those who are unfamiliar with the libraries.
- [Setup Software for Non-RPi Computers.txt](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Setup%20Software%20for%20Non-RPi%20Computers.txt) -  Information on how to get the GUI running on a non-RaspberryPi computer for software development.
- [Software Coding Standard.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Software%20Coding%20Standard.docx) -	The Coding Standard for coding used throughout the software.
- [Software Update Procedure.md](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Software%20Update%20Procedure.md) - 	Information on how to utilize GitHub when updating the on-device software.
- Supplementary Directory:
  - [Kivy Clock Summary.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Supplementary/Kivy%20Clock%20Summary.docx) - 	Information recorded by the capstone team about the logic for using the clock the way they did.
  - [Raspberry Pi Setup.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Supplementary/Raspberry%20Pi%20Setup.docx) - 	Information on how to set up a new raspberry pi from scratch to host the software.
### Examples
- [Example_Adding the Camera Module.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Example_Adding%20the%20Camera%20Module.docx) - A reflection on how the camera screens were added.
- [Example_Adding a Boot Screen.docx](https://github.com/byu-crop-biomechanics-lab/FIELDAQ/blob/master/Documentation/Example_Adding%20a%20Boot%20Screen.docx) - A hypothetical situation where a boot screen is created to choose between the Camera test mode and the Stalk Push testing mode.

## Plans for Future Changes and Improvements
- There are a few GitHub branches of the repository in various states of progress. Their subjects of feature improvements include:
  - Implementing a bar code scanner for naming test files.
  - Adjusting the camera settings
  - Fixing an issue with the image review screens where some images may not appear.
  - Creating a new keyboard layout with a minus sign to be used for text entry in the calibration screens.
- A future redesign of the PCB would be convenient for adjusting port placement on the physical board.

## Notes from Maize Runners (Capstone Team 7, 2022-2023)
- A list of changes made for the ARM device
Replaced Potentiometer Angle sensor with Friction Load cell sensor 

Files added: 

FIELDAQ/Granusoft/src/sensors/Friction_Load.py 

Files changed: 

FIELDAQ/Granusoft/src/Sensor/SensorReal.py 

FIELDAQ/Granusoft/src/Sensor/SensorFake.py 

FIELDAQ/Granusoft/src/sensors/connections.py 

FIELDAQ/Granusoft/src/Dataset.py 

FIELDAQ/Granusoft/src/view/screens/main/LiveFeedScreen.kv 

FIELDAQ/Granusoft/src/view/screens/main/LiveFeedScreen.py 

FIELDAQ/Granusoft/src/view/screens/main/testing/SaveScreen.py 

FIELDAQ/Granusoft/src/view/screens/main/testing/TestDetailScreen.py 

FIELDAQ/Granusoft/src/view/screens/main/testing/TestingResultsScreen.py 

FIELDAQ/Granusoft/src/view/screens/main/testing/TestInProgressScreen.kv 

FIELDAQ/Granusoft/src/view/screens/main/testing/TestInProgressScreen.py 

 

Added limit switch to sensors and dataset 

Files changed: 

FIELDAQ/Granusoft/src/Dataset.py 

FIELDAQ/Granusoft/src/view/screens/main/testing/SaveScreen.py 

FIELDAQ/Granusoft/src/view/screens/main/testing/TestInProgressScreen.py 

 

Changed lbs to N (newtons) 

Files changed: 

FIELDAQ/Granusoft/src/view/screens/main/testing/TestingResultsScreen.kv 

FIELDAQ/Granusoft/src/view/screens/main/testing/TestDetailScreen.kv 

FIELDAQ/Granusoft/src/view/screens/main/testing/TestInProgressScreen.kv 

FIELDAQ/Granusoft/src/view/screens/main/LiveFeedScreen.py 


src/history.txt
    1  lsusb
    2  cd ~
    3  ls
    4  cd granusoft
    5  cd fieldaq
    6  cd GranuSoft/
    7  ls
    8  sudo mount -t vfat -o uid=pi,gid=pi /devsda1 /mnt/usbStick
    9  sudo mount -t vfat -o uid=pi,gid=pi /dev/sda1 /mnt/usbStick
   10  sudo cp -r /mnt/usbStick/FIELDAQ ~/
   11  sudo umount /mnt/usbStick
   12  python3 main.py
   13  cd ..
   14  cd FIELDAQ/Granusoft/src/
   15  python3 main.py 
   16  reboot
   17  sudo mount -t vfat -o uid=pi,gid=pi /devsda1 /mnt/usbStick
   18  sudo mount -t vfat -o uid=pi,gid=pi /dev/sda1 /mnt/usbStick
   19  sudo cp -r /mnt/usbStick/FIELDAQ ~/
   20  sudo umount /mnt/usbStick
   21  python3 main.py 
   22  sudo mount -t vfat -o uid=pi,gid=pi /dev/sda1 /mnt/usbStick
   23  sudo cp -r /mnt/usbStick/FIELDAQ ~/
   24  sudo umount /mnt/usbStick
   25  python3 main.py 
   26  sudo mount -t vfat -o uid=pi,gid=pi /dev/sda1 /mnt/usbStick
   27  sudo cp -r /mnt/usbStick/FIELDAQ ~/
   28  python3 main.py 
   29  sudo mount -t vfat -o uid=pi,gid=pi /dev/sda1 /mnt/usbStick
   30  sudo umount /mnt/usbStick
   31  python3 main.py 
   32  ping 8.8.8.8
   33  pip3 install pytz
   34  ping 8.8.8.8
   35  sudo mount -t vfat -o uid=pi,gid=pi /dev/sda1 /mnt/usbStick
   36  sudo cp -r /mnt/usbStick/FIELDAQ ~/
   37  sudo umount /mnt/usbStick
   38  python3 main.py 
   39  sudo mount -t vfat -o uid=pi,gid=pi /dev/sda1 /mnt/usbStick
   40  sudo cp -r /mnt/usbStick/FIELDAQ ~/
   41  sudo umount /mnt/usbStick
   42  python3 main.py 
   43  sudo mount -t vfat -o uid=pi,gid=pi /dev/sda1 /mnt/usbStick
   44  sudo cp -r /mnt/usbStick/FIELDAQ ~/
   45  sudo umount /mnt/usbStick
   46  python3 main.py 
   47  exit
   48  ifconfig
   49  iwconfig
   50  hostname -I
   51  ip r | grep default
   52  sudo cat /etc/resolv.conf
   53  sudo nano /etc/dhcpcd.conf
   54  sudo reboot
   55  ifconfig
   56  sudo nano /etc/resolv.conf
   57  ip r | grep default
   58  sudo nano /etc/dhcpcd.conf
   59  hostname -I
   60  sudo nano /etc/resolv.conf
   61  sudo nano /etc/dhcpcd.conf
   62  ls
   63  ll
   64  ls -l
   65  sudo reboot
   66  exit
   67  ls
   68  cd Tests
   69  ls
   70  cd jordan
   71  ls
   72  cd Tests_2022_06_22/
   73  ls
   74  cd ..
   75  pwd
   76  exit
   77  ls
   78  cd tests
   79  cd Tests
   80  ls
   81  cd jordan
   82  ls
   83  exit
   84  ls
   85  cd Tests
   86  ls
   87  cd jordan
   88  ls
   89  exit
   90  ls
   91  cd Tests
   92  ls
   93  cd jordan
   94  ls
   95  exit
   96  ls
   97  cd Tests
   98  ls
   99  cd jordan
  100  ls
  101  ls -l
  102  pwd
  103  exit
  104  cd Tests
  105  ls
  106  cd jordan
  107  ls
  108  cat 2023_01_13_09_16_10_P75_T36.csv 
  109  cd Tests
  110  cd jordan
  111  ls
  112  exit
  113  cd Tests/jordan/
  114  ls
  115  exit
  116  cd Tests
  117  cd jordan
  118  ls
  119  exit
  120  ls
  121  cd Tests
  122  ls
  123  cd jordan
  124  ls
  125  pwd
  126  exit
  127  cd Tests/jordan
  128  ls
  129  exit
  130  cd Tests
  131  cd jordan
  132  ls
  133  cat 2023_01_17_08_34_08_P75_T41.csv 
  134  df -H
  135  cat 2023_01_17_08_34_08_P75_T41.csv 
  136  ls
  137  cat 2023_01_17_08_41_11_P75_T42.csv 
  138  ls
  139  pwd
  140  exit
  141  git status
  142  ls
  143  cd sensors
  144  ls
  145  cat connections.py 
  146  code connections.py 
  147  git status
  148  cd ..
  149  code .
  150  ifconfig
  151  ifconfig
  152  reboot
  153  history
  154  touch history.txt
  155  ls
  156  doskey /history > history.txt 
  157  history > history.txt 


src/sensors/connections.py
import time
import board
import busio
import digitalio
from time import sleep
import RPi.GPIO as GPIO
import serial
import adafruit_gps
import adafruit_lis3dh
#import adafruit_am2320
import adafruit_ads1x15.ads1115 as ADS
from adafruit_ads1x15.analog_in import AnalogIn

GPIO.setmode(GPIO.BCM)
i2c = busio.I2C(board.SCL, board.SDA, 115200)
uart = serial.Serial("/dev/ttyS0", baudrate=9600, timeout=3000)

# Temperature and Humidity sensor, off of the I2C pins on bottom right of board
#am = adafruit_am2320.AM2320(i2c)

# Accelerometer, top middle of board
# lis3dh = adafruit_lis3dh.LIS3DH_I2C(i2c)
# lis3dh.range = adafruit_lis3dh.RANGE_2_G


# GPIO, right edge of board
GPIO1 = 4 #BOARD 7, BCM 4
GPIO2 = 17 #BOARD 11, BCM 17
GPIO3 = 27 #BOARD 13, BCM 27
GPIO4 = 22 #BOARD 15, BCM 22

# SPI/GPIO, top edge of board
SPI_CE1 = 7 #BOARD 26, BCM 7
SPI_CE0 = 8 #BOARD 24, BCM 8
SPI_SCLK = 11 #BOARD 23, BCM 11
SPI_MISO = 9 #BOARD 21, BCM 9
SPI_MOSI = 10 #BOARD 19, BCM 10

GPIO_PINS = [GPIO1, GPIO2, GPIO3, GPIO4, SPI_CE1, SPI_CE0,\
        SPI_SCLK, SPI_MISO, SPI_MOSI]

# for pin in GPIO_PINS:
    # GPIO.setup(pin, GPIO.OUT) #change to out for every single one except the limit switch pins

GPIO.setup(GPIO1, GPIO.IN)
GPIO.setup(GPIO2, GPIO.IN)
GPIO.setup(GPIO3, GPIO.IN)
GPIO.setup(GPIO4, GPIO.IN)
GPIO.setup(SPI_CE1, GPIO.IN)
GPIO.setup(SPI_CE0, GPIO.IN)
GPIO.setup(SPI_SCLK, GPIO.IN)
GPIO.setup(SPI_MISO, GPIO.IN)
GPIO.setup(SPI_MOSI, GPIO.IN)


# ADC
ads1 = ADS.ADS1115(i2c, address=0x49, data_rate = 860, mode=0)
ads2 = ADS.ADS1115(i2c, address=0x48, data_rate = 860, mode=0, gain=2)
CHAN0 = AnalogIn(ads1, ADS.P0)
CHAN1 = AnalogIn(ads1, ADS.P1)
CHAN2 = AnalogIn(ads1, ADS.P2)
CHAN3 = AnalogIn(ads1, ADS.P3)
CHAN4 = AnalogIn(ads2, ADS.P0) # farthest from vcc and ground on pinout connector
CHAN5 = AnalogIn(ads2, ADS.P1)
CHAN6 = AnalogIn(ads2, ADS.P2)
CHAN7 = AnalogIn(ads2, ADS.P3) # closest to vcc and ground on pinout connector

# Channels for the pot and force sensors

#HEIGHT_POT_CHAN = CHAN1
FRICTION_CHAN = CHAN2

X_LOAD_CHAN = CHAN4
#Y_LOAD_CHAN = CHAN2

# Scaling factor for the force sensor
FORCE_SENSOR_SCALING = 16 # 3556.1878
# ads2.gain = FORCE_SENSOR_SCALING

'''
# GPS
gps = adafruit_gps.GPS(uart, debug=False)
gps.send_command(b'PMTK314,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0')
gps.send_command(b'PMTK220,500')
try:
    gps.update()
except:
    print("GPS may have a problem. Try Rebooting")

# MOTORS
PWMA = 16 #BOARD 36, BCM 16
AIN1 = 25 #BOARD 22, BCM 25
AIN2 = 20 #BOARD 38, BCM 20

PWMB = 21 #BOARD 40, BCM 21
BIN1 = 5 #BOARD 29, BCM 5
BIN2 = 12 #BOARD 32, BCM 12

PWMC = 13 #BOARD 33, BCM 13
CIN1 = 26 #BOARD 37, BCM 26
CIN2 = 19 #BOARD 35, BCM 19

PWMD = 24 #BOARD 18, BCM 24
DIN1 = 18 #BOARD 12, BCM 18
DIN2 = 23 #BOARD 16, BCM 23

MOTORS = ['A', 'B', 'C', 'D']
IN1 = [AIN1, BIN1, CIN1, DIN1]
IN2 = [AIN2, BIN2, CIN2, DIN2]
PWM = [PWMA, PWMB, PWMC, PWMD]

for in1, in2, pwm in zip(IN1, IN2, PWM):
    GPIO.setup(in1, GPIO.OUT)
    GPIO.setup(in2, GPIO.OUT)
    GPIO.setup(pwm, GPIO.OUT)
'''
